# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [8]:
import pandas as pd

df = pd.read_csv("../data/2019_Yellow_Taxi_Trip_Data.csv")
df.head(n=5)

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5    1.0      0.5   
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [9]:
df.shape

(10000, 18)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [20]:
df[["fare_amount", "tip_amount", "tolls_amount", "total_amount"]].describe(include='all')


fare_amount    tip_amount  tolls_amount  total_amount
count  10000.000000  10000.000000  10000.000000  10000.000000
mean      15.106313      2.634494      0.623447     22.564659
std       13.954762      3.409800      6.437507     19.209255
min      -52.000000      0.000000     -6.120000    -65.920000
25%        7.000000      0.000000      0.000000     12.375000
50%       10.000000      2.000000      0.000000     16.300000
75%       16.000000      3.250000      0.000000     22.880000
max      176.000000     43.000000    612.000000    671.800000

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [19]:
df.loc[df["trip_distance"].idxmax(), ["fare_amount", "tip_amount", "tolls_amount", "total_amount"]]

fare_amount      176.0
tip_amount       18.29
tolls_amount      6.12
total_amount    201.21
Name: 8338, dtype: object

---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [35]:
df = pd.read_csv("../data/Meteorite_Landings.csv")

df = df.rename(
    columns={"mass (g)": "mass"}
)

#df.columns
#df.head()
df = df.drop(columns=["GeoLocation", "reclat", "reclong"])

df.sort_values("mass", ascending=False)

name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   
...                ...    ...      ...           ...         ...    ...   
38282   Wei-hui-fu (a)  24231    Valid          Iron         NaN  Found   
38283   Wei-hui-fu (b)  24232    Valid          Iron         NaN  Found   
38285          Weiyuan  24233    Valid  Mesosiderite         NaN  Found   
41472    Yamato 792768  28117    Valid           CM2         NaN  Found   
45698    Zapata County  30393    Valid          Iron         NaN  Found   

                         year  
16392  01/01/1920 12:00:00 AM  
5373   01/01/1818 12:00:00 AM  
5365   12/22/1575 12:00:00 AM  
5370   01/01/1891 12:00:00 AM  
3455   01/01/1898 12:00:00 AM  
...                       ...  
38282  01/01/1931 12:00:00 AM  
38283  01/01/1931 12:00:00 AM  
38285  01/01/1978 12:00:00 AM  
41472  01/01/1979 12:00:00 AM  
45698  01/01/1930 12:00:00 AM  

[45716 rows x 7 columns]

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [32]:
df = pd.read_csv("../data/Meteorite_Landings.csv")
df["fall"].unique()

array(['Fell', 'Found'], dtype=object)

In [34]:
df = pd.read_csv("../data/Meteorite_Landings.csv").assign(
    year=lambda x: pd.to_numeric(x.year.str.slice(6,10)),
    pre_1970=lambda x: (x.fall=="Fell") & (x.year < 1970),
)

df = df.set_index('id').sort_index(ascending=True)
df.loc[10036:10040]

name nametype         recclass  mass (g)   fall    year    reclat  \
id                                                                              
10036     Enigma    Valid               H4      94.0  Found  1967.0  31.33333   
10037       Enon    Valid  Iron, ungrouped     763.0  Found  1883.0  39.86667   
10038      Enshi    Valid               H5    8000.0   Fell  1974.0  30.30000   
10039  Ensisheim    Valid              LL6  127000.0   Fell  1491.0  47.86667   

         reclong            GeoLocation  pre_1970  
id                                                 
10036  -82.31667  (31.33333, -82.31667)     False  
10037  -83.95000     (39.86667, -83.95)     False  
10038  109.50000          (30.3, 109.5)     False  
10039    7.35000       (47.86667, 7.35)      True

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

In [42]:
meteorites = pd.read_csv("../data/Meteorite_Landings.csv").assign(year=lambda x: pd.to_numeric(x.year.str.slice(6,10)))
meteorites.head()
pd.set_option('display.float_format', '{:,.0f}'.format)
meteorites.query('year.between(2005, 2009)').pivot_table(
    index='year', columns='fall', values='mass (g)',
    aggfunc=['count', lambda x: x.quantile(0.95)]
).rename(columns=
         {'<lambda>': '95 percentile'}
        )

count       95 percentile      
fall   Fell Found          Fell Found
year                                 
2,005   NaN   874           NaN 4,500
2,006     5 2,450        25,008 1,600
2,007     8 1,181        89,675 1,127
2,008     9   948       106,000 2,275
2,009     5 1,492         8,333 1,397

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

In [44]:
meteorites = pd.read_csv("../data/Meteorite_Landings.csv")
meteorites.groupby('fall')['mass (g)'].describe()

count   mean     std  min  25%   50%    75%        max
fall                                                         
Fell   1,075 47,071 717,067    0  686 2,800 10,450 23,000,000
Found 44,510 12,462 571,106    0    7    30    178 60,000,000

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

In [5]:
import pandas as pd
taxis = pd.read_csv("../data/2019_Yellow_Taxi_Trip_Data.csv")
taxis.dtypes

vendorid                   int64
tpep_pickup_datetime      object
tpep_dropoff_datetime     object
passenger_count            int64
trip_distance            float64
ratecodeid                 int64
store_and_fwd_flag        object
pulocationid               int64
dolocationid               int64
payment_type               int64
fare_amount              float64
extra                    float64
mta_tax                  float64
tip_amount               float64
tolls_amount             float64
improvement_surcharge    float64
total_amount             float64
congestion_surcharge     float64
dtype: object

In [16]:
import pandas as pd
taxis = pd.read_csv("../data/2019_Yellow_Taxi_Trip_Data.csv", parse_dates=["tpep_dropoff_datetime"])

taxis = taxis.set_index("tpep_dropoff_datetime")
taxis = taxis.sort_index()
taxis_reasmpled = taxis.resample('1h')[[
    "trip_distance", "fare_amount", "tolls_amount", "tip_amount"
    ]].agg('sum').nlargest(5, "tip_amount")
print(taxis_reasmpled)

                       trip_distance  fare_amount  tolls_amount  tip_amount
tpep_dropoff_datetime                                                      
2019-10-23 16:00:00         10676.95     67797.76        699.04    12228.64
2019-10-23 17:00:00         16052.83     70131.91       4044.04    12044.03
2019-10-23 18:00:00          3104.56     11565.56       1454.67     1907.64
2019-10-23 15:00:00            14.34       213.50          0.00       51.75
2019-10-23 19:00:00            98.59       268.00         24.48       25.74


--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.